### ISSUE TITLE
Wrong non-redirect behaviour with "null" path info

### ISSUE THREAD

**Issue by TobiX** [eclipse/jetty.project#3023]
When using the `DefaultServlet` to serve sub-directories and `pathInfoOnly` set to `true`, the topmost directory isn't redirected from "path without slash" to "path with slash", but serves the welcome file or directory index directly.

It seems there was a special case for this in the dark history in Jetty using the `org.eclipse.jetty.server.nullPathInfo` attribute, but I couldn't track down where this was ever set:

https://github.com/eclipse/jetty.project/blob/cd0cf3d11f139fb068321a0c4ef56ff1c3cdf087/jetty-server/src/main/java/org/eclipse/jetty/server/ResourceService.java#L376

Here is a sample class with a simple workaround:
````java
import java.io.IOException;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.eclipse.jetty.http.HttpContent;
import org.eclipse.jetty.server.ResourceService;
import org.eclipse.jetty.server.Server;
import org.eclipse.jetty.servlet.DefaultServlet;
import org.eclipse.jetty.servlet.ServletContextHandler;
import org.eclipse.jetty.servlet.ServletHolder;

public class TestServer {

  public static void main (String[] args) throws Exception {
    Server jetty = new Server (1234);
    ServletContextHandler handler = new ServletContextHandler ();
    handler.setContextPath ("/");
    DefaultServlet defaultServlet = new DefaultServlet (new FixedResourceService ());
    ServletHolder holder = new ServletHolder ("default", defaultServlet);
    holder.setInitParameter ("pathInfoOnly", "true");
    holder.setInitParameter ("resourceBase", ".");
    handler.addServlet (holder, "/test/*");
    jetty.setHandler (handler);
    jetty.start ();
  }

  private static class FixedResourceService extends ResourceService {

    @Override
    protected void sendWelcome (HttpContent content,
                                String pathInContext,
                                boolean endsWithSlash,
                                boolean included,
                                HttpServletRequest request,
                                HttpServletResponse response) throws ServletException, IOException {
      if (request.getPathInfo () == null) {
        request.setAttribute ("org.eclipse.jetty.server.nullPathInfo", true);
      }
      super.sendWelcome (content, pathInContext, endsWithSlash, included, request, response);
    }
  }

}
````

I think just checking for `request.getPathInfo () == null` directly in `sendWelcome` (instead of the check for the non-existing attribute) would fix this problem.

**Comment by eicki** [comment]
The issue has probably been since
https://github.com/eclipse/jetty.project/commit/564ffca5dceb9bd54388770533a0f6065a5583ca
which added a tiny _contextPath.length() > 1 in the ContextHandler
{code}
// Are we not the root context?
// redirect null path infos
if (!_allowNullPathInfo && _contextPath.length() == target.length() && _contextPath.length() > 1)
{code}

**Comment by joakime** [comment]
Is your declared `DefaultServlet` the only one you are using?
You should declare the Resource Base on the on the `ServletContextHandler` (otherwise many important `javax.servlet.ServletContext` methods will not work).
There should always be a servlet mapped to `/`, typically you want the normal (servlet spec default) `DefaultServlet` (otherwise its a 404)
If you declare a new `DefaultServlet` and use it for alternative static file content, you should not use relative paths (especially just `"."`) on your `resourceBase` init-param as your resulting Resource will be ambiguous (is it a File System? or a URL? or Jar context?) and possibly be different on each run, use either a fully qualified path (which results in a PathResource), or an absolute URL (which results in a URLResource).

Example using both a "default" `DefaultServlet` and a custom one in the same `ServletContext` :

``` java
package jetty;

import java.io.File;
import java.nio.file.Path;

import org.eclipse.jetty.server.Server;
import org.eclipse.jetty.server.ServerConnector;
import org.eclipse.jetty.server.handler.DefaultHandler;
import org.eclipse.jetty.server.handler.HandlerList;
import org.eclipse.jetty.servlet.DefaultServlet;
import org.eclipse.jetty.servlet.ServletContextHandler;
import org.eclipse.jetty.servlet.ServletHolder;
import org.eclipse.jetty.util.resource.PathResource;

public class ManyDefaultServlet
{
    public static void main(String[] args) throws Exception {
        Server server = new Server();
        ServerConnector connector = new ServerConnector(server);
        connector.setPort(8080);
        server.addConnector(connector);

        // The filesystem paths we will map
        Path homePath = new File(System.getProperty("user.home")).toPath().toRealPath();
        Path pwdPath = new File(System.getProperty("user.dir")).toPath().toRealPath();
        
        // Setup the basic application "context" for this application at "/"
        // This provides the javax.servlet.ServletContext and associated behaviors
        ServletContextHandler context = new ServletContextHandler(ServletContextHandler.SESSIONS);
        context.setContextPath("/");
        context.setWelcomeFiles(new String[] {
                // Use path specific html files
                "index.html", 
                "index.htm", 
                "welcome.html",
                // Use root /index.html if path specific one doesn't exist
                "/index.html"});
        
        // Establish a Base Resource for ServletContext, otherwise the following methods will not work
        //  - Set<String> getResourcePaths(String)
        //  - URL getResource(String)
        //  - InputStream getResourceAsStream(String)
        //  - String getRealPath(String)
        context.setBaseResource(new PathResource(pwdPath));

        // Fist, add special pathspec of "/home/" content mapped to the homePath
        ServletHolder holderHome = new ServletHolder("static-home", DefaultServlet.class);
        // This should be a fully qualified Path or absolute URI (not relative directory)
        holderHome.setInitParameter("resourceBase",homePath.toUri().toASCIIString()); 
        holderHome.setInitParameter("dirAllowed","true");
        // Use request pathInfo, don't calculate from contextPath
        holderHome.setInitParameter("pathInfoOnly","true");
        context.addServlet(holderHome,"/home/*"); // must end in "/*" for pathInfo to work

        // Lastly, the default servlet for root content
        // It is important that this is last.
        String defName = "default"; // the important "default" name
        ServletHolder holderDef = new ServletHolder(defName, DefaultServlet.class);
        holderDef.setInitParameter("dirAllowed","true"); // tweak it some
        context.addServlet(holderDef,"/"); // the servlet spec "default url-pattern"

        HandlerList handlers = new HandlerList();
        handlers.addHandler(context); // the context
        handlers.addHandler(new DefaultHandler()); // for producing errors if no above handlers match
        server.setHandler(handlers);
        
        server.start();
        server.join();
    }
}
```

**Comment by joakime** [comment]
Closing as fix was merged into `jetty-9.4.x` branch on commit 3995cd2ec0ffb6f756507ad99295bfced287ba0b
